docs/spec/lsm/adr-002.md

```diff
@@ -0,0 +1,362 @@
+```
+
+### Helper functions
+In order to identify whether a liquid stake transaction will exceed either the global liquid staking cap or the validator bond cap, the following functions were added:
+
+```go
+// Check if an account is a owned by a liquid staking provider
+// This is determined by checking if the account is a 32-length module account
+func (k Keeper) AccountIsLiquidStakingProvider(address sdk.AccAddress) bool 
+
+// SafelyIncreaseTotalLiquidStakedTokens increments the total liquid staked tokens
+// if the caps are enabled and the global cap is not surpassed by this delegation
+func (k Keeper) SafelyIncreaseTotalLiquidStakedTokens(ctx sdk.Context, amount sdk.Int) error 
+
+// DecreaseTotalLiquidStakedTokens decrements the total liquid staked tokens
+// if the caps are enabled
+func (k Keeper) DecreaseTotalLiquidStakedTokens(ctx sdk.Context, amount sdk.Int) error
+
+// SafelyIncreaseValidatorTotalLiquidShares increments the total liquid shares on a validator
+// if the caps are enabled and the validator bond cap is not surpassed by this delegation
+func (k Keeper) SafelyIncreaseValidatorTotalLiquidShares(ctx sdk.Context, validator types.Validator, shares sdk.Dec) error 
+
+// DecreaseValidatorTotalLiquidShares decrements the total liquid shares on a validator
+// if the caps are enabled
+func (k Keeper) DecreaseValidatorTotalLiquidShares(ctx sdk.Context, validator types.Validator, shares sdk.Dec) error
+
+// SafelyDecreaseValidatorBond decrements the total validator's self bond
+// so long as it will not cause the current delegations to exceed the threshold
+// set by validator bond factor
+func (k Keeper) SafelyDecreaseValidatorBond(ctx sdk.Context, validator types.Validator, shares sdk.Dec) error 
+```
+
+### Accounting
+Tracking the total liquid stake and total liquid validator shares requires additional accounting changes in the following transactions/events:
+
+```go
+func Delegate() {
+    ...
+    // If delegator is a liquid staking provider
+    //    Increment total liquid staked
+    //    Increment validator liquid shares
+}
+
+func Undelegate() {
+    ...
+    // If delegator is a liquid staking provider
+    //    Decrement total liquid staked
+    //    Decrement validator liquid shares
+}
+
+func BeginRedelegate() {
+    ...
+    // If delegator is a liquid staking provider
+    //    Decrement source validator liquid shares
+    //    Increment destination validator liquid shares
+}
+
+func TokenizeShares() {
+    ...
+    // If delegator is a NOT liquid staking provider (otherwise the shares are already included)
+    //    Increment total liquid staked
+    //    Increment validator liquid shares
+}
+
+func RedeemTokens() {
+    ...
+    // If delegator is a NOT liquid staking provider 
+    //    Decrement total liquid staked
+    //    Decrement validator liquid shares
+}
+
+func Slash() {
+    ...
+    // Decrement total liquid staked (since total liquid stake is denominated in tokens, not shares)
```

Comment:
This needs to be better explained as it's not trivial.

Reply:
added clarification [here]([added clarification around slash accounting](https://github.com/cosmos/cosmos-sdk/pull/16747/commits/02a42df5c600109161003adc47d60bc8e0f4a00f))

Reply:
@mpoke can I mark this resolved?